### PR TITLE
Selam Ainalem Amperes

### PR DESCRIPTION
**Pull request by SelamawitA**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | Pseudocding helped sort out my design decisions, I attempted to rely on previously made methods to reduce the amount of lines in my homework and taking the time to write before coding made that process easier. 
Describe any examples of composition that you encountered in this project, if any | Composition was used in the passing of trip objects into instantiation of the passenger and driver objects. 
Describe a _nominal test_ that you wrote for this assignment. | I created a test for request_trip to ensure that I was able to receive a Trip object. 
Describe an _edge case test_ that you wrote for this assignment. | I made my request_trip test case more complex by adding multiple trips and looking into the attributes that the trip object produced. For example, ensuring that the driver's status was changed to unavailable and that the trip array was updated after the method had run. 
Describe a concept that you gained more clarity on as you worked on this assignment. | I feel more comfortable with the use of composition. I really liked how we were shown a different way composition can be used and that the methods all relied on each other instead of building on each other like I've seen in most examples.

### CODE REVIEW

**Comment by tildeee** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | x
Answer comprehension questions | x
**Wave 1** |
`Trip` has a helper method to calculate duration | yes, but no test for this
`Passenger` has a method to calculate total cost of all trips | x
`Driver` has a method to calculate total revenue | x
Complete tests | most nominal cases, notes below
**Wave 2** |
`TripDispatcher` has a new method to create trips | x
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | You do `a_passenger.trips << new_trip`. It would be better if you invoked a method off of passenger, such as `a_passenger.add_trip(new_trip)`
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | untested
Complete tests | nominally, yes
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | the tests for this are weird, I'm adding some comments about it below!
Appropriate use of Ruby's `Date` | x
Appropriate helper methods were made to help with complex logic | you ended up choosing to make really long variable names instead of creating separate helper methods
Complete tests | 
**Overall** | 

I think you did a really great job of doing complex things with enumerables and iteration! Great job

You got most of the nominal cases for these functionalities, but missed some edge cases:

You could have added the following edge cases:
- `Passenger` `total cost` method for passengers with no trips
- `Passenger` `all time spent` method for passengers with no trips
- `Passenger` methods for trips with a `nil` end_time

I'm glad you got through wave 3. In general, your tests around making a new test and assigning a driver are a little strange and off, so you'll see some comments on that below.

No red flags from me though, but I want to make sure that when we start practicing complex logic in rails, you nail the testing on it! :sparkles:

File: specs/trip_dispatch_spec.rb [hunk 1]
```diff
@@ -17,7 +20,85 @@
...
+        if a_driver.status == :AVAILABLE && a_driver.name != new_trip_for_longest_wait.driver.name
+          a_driver.trips.length.must_be :>, length_of_longest_wait - a_single_trip
+        end
+      end
+
+      length_of_longest_wait.must_equal 1
+    end
+
+    it "can send the driver who has waited the longest to a passenger first (excluding new drivers from test)" do
+      first_driver_end_time = Time.parse('2017-02-19 18:39:00 +0000')
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:75]: this variable first_driver_end_time ends up not being used?

File: specs/trip_dispatch_spec.rb [hunk 2]
```diff
@@ -17,7 +20,85 @@
...
+      last_driver_end_time = Time.parse('2017-01-28 13:01:00 +0000')
+
+      dispatcher = RideShare::TripDispatcher.new
+      all_available_drivers = dispatcher.drivers.find_all{|driver|driver.status == :AVAILABLE}
+      passenger = 1
+      (all_available_drivers.length-2).times do
+        dispatcher.assign_by_driver_status(passenger)
+        passenger+=1
+      end
+      reduced_available_drivers = dispatcher.drivers.find_all{|driver|driver.status == :AVAILABLE}
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:85]: `reduced_available_drivers` is a local variable never used, and I'm not sure what you're trying to do with it!

File: specs/trip_dispatch_spec.rb [hunk 3]
```diff
@@ -17,7 +20,85 @@
...
+      new_trip = dispatcher.assign_by_driver_status(46)
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:86]: Where does 46 come from?

File: specs/trip_dispatch_spec.rb [hunk 4]
```diff
@@ -17,7 +20,85 @@
...
+      new_trip.driver.trips[-2].end_time.must_equal last_driver_end_time
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:87]: Where does -2 come from?

File: specs/trip_dispatch_spec.rb [hunk 5]
```diff
@@ -17,7 +20,85 @@
...
+    it "can send the driver who has waited the longest to a passenger first (excluding new drivers from test)" do
+      first_driver_end_time = Time.parse('2017-02-19 18:39:00 +0000')
+      last_driver_end_time = Time.parse('2017-01-28 13:01:00 +0000')
+
+      dispatcher = RideShare::TripDispatcher.new
+      all_available_drivers = dispatcher.drivers.find_all{|driver|driver.status == :AVAILABLE}
+      passenger = 1
+      (all_available_drivers.length-2).times do
+        dispatcher.assign_by_driver_status(passenger)
+        passenger+=1
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:83]: Why does passenger need to increase by 1?

File: specs/trip_dispatch_spec.rb (hunk 4 above)
Comment (tildeee) [reply specs/trip_dispatch_spec.rb:87]: Does asking and just checking equality on `end_time` give you confidence that this method works?

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -17,7 +20,85 @@
...
+      new_trip_for_longest_wait = dispatcher.assign_by_driver_status(1)
+      length_of_longest_wait = new_trip_for_longest_wait.driver.trips.length
+
+      dispatcher.drivers.each do |a_driver|
+        if a_driver.status == :AVAILABLE && a_driver.name != new_trip_for_longest_wait.driver.name
+          a_driver.trips.length.must_be :>, length_of_longest_wait - a_single_trip
+        end
+      end
+
+      length_of_longest_wait.must_equal 1
```
Comment (tildeee) [inline specs/trip_dispatch_spec.rb:71]: Does checking that the length of longest wait equals 1 give you confidence that this method works?
